## <b> The Play Store apps data has enormous potential to drive app-making businesses to success. Actionable insights can be drawn for developers to work on and capture the Android market. </b>

## <b> Each app (row) has values for catergory, rating, size, and more. Another dataset contains customer reviews of the android apps.</b>

## <b> Explore and analyze the data to discover key factors responsible for app engagement and success. </b>

In [ ]:
# Import pandas package
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# loading csv File
file_path = '/content/drive/MyDrive/Datasets/data/'
df_PlayStoreData = pd.read_csv(file_path+ 'Play Store Data.csv')

In [ ]:
# loading csv File
file_path = '/content/drive/MyDrive/Datasets/data/EDA-Play store apps/User Reviews.csv'
df_User_Review = pd.read_csv(file_path)

**Exploring the Data**

In [ ]:
#Data Exploration
df_PlayStoreData.head(3)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

  Reviews  Size    Installs  Type Price Content Rating  \
0     159   19M     10,000+  Free     0       Everyone   
1     967   14M    500,000+  Free     0       Everyone   
2   87510  8.7M  5,000,000+  Free     0       Everyone   

                      Genres      Last Updated Current Ver   Android Ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up  
2               Art & Design    August 1, 2018       1.2.4  4.0.3 and up

In [ ]:
df_User_Review.head(50)

App                                  Translated_Review  \
0   10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1   10 Best Foods for You    This help eating healthy exercise regular basis   
2   10 Best Foods for You                                                NaN   
3   10 Best Foods for You         Works great especially going grocery store   
4   10 Best Foods for You                                       Best idea us   
5   10 Best Foods for You                                           Best way   
6   10 Best Foods for You                                            Amazing   
7   10 Best Foods for You                                                NaN   
8   10 Best Foods for You                               Looking forward app,   
9   10 Best Foods for You              It helpful site ! It help foods get !   
10  10 Best Foods for You                                          good you.   
11  10 Best Foods for You  Useful information The amount spelling errors ...   
12  10 Best Foods for You  Thank you! Great app!! Add arthritis, eyes, im...   
13  10 Best Foods for You  Greatest ever Completely awesome maintain heal...   
14  10 Best Foods for You  Good health...... Good health first priority.....   
15  10 Best Foods for You                                                NaN   
16  10 Best Foods for You  Health It's important world either life . thin...   
17  10 Best Foods for You  Mrs sunita bhati I thankful developers,to make...   
18  10 Best Foods for You  Very Useful in diabetes age 30. I need control...   
19  10 Best Foods for You                                 One greatest apps.   
20  10 Best Foods for You                                          good nice   
21  10 Best Foods for You                              Healthy Really helped   
22  10 Best Foods for You                                         God health   
23  10 Best Foods for You  HEALTH SHOULD ALWAYS BE TOP PRIORITY. !!. ON M...   
24  10 Best Foods for You                              An excellent A useful   
25  10 Best Foods for You                  I found lot wealth form health...   
26  10 Best Foods for You                         Because I found important.   
27  10 Best Foods for You                                     Healthy Eating   
28  10 Best Foods for You                              Very good Simply good   
29  10 Best Foods for You                                        On test....   
30  10 Best Foods for You                                            Good.!!   
31  10 Best Foods for You  Thanks advice. Downloaded Adobe reader still c...   
32  10 Best Foods for You                 No recipe book Unable recipe book.   
33  10 Best Foods for You                     Absolutely Fabulous Phenomenal   
34  10 Best Foods for You                                    Great Love food   
35  10 Best Foods for You                          Excellent It really works   
36  10 Best Foods for You                                               Best   
37  10 Best Foods for You  WEALTH OF INFORMATION Very informative ... eas...   
38  10 Best Foods for You                               Not description good   
39  10 Best Foods for You                                    Great app. Love   
40  10 Best Foods for You                                               Good   
41  10 Best Foods for You                                 Loved Very helpful   
42  10 Best Foods for You                                Good healthy foods.   
43  10 Best Foods for You  Waste time It needs internet time n ask calls ...   
44  10 Best Foods for You    Best thing I lost ten lbs. two months help tips   
45  10 Best Foods for You  Ten best food promote good health Good health ...   
46  10 Best Foods for You   Healthy Food Really good information eat healthy   
47  10 Best Foods for You         My mom dad important thing This good share   
48  10 Best Foods for You                A big thanks ds I got bst gd health   
49  10 Best Foods for 

In [ ]:
#Data Exploration
df_PlayStoreData.tail(3)

App             Category  \
10838                         Parkinson Exercices FR              MEDICAL   
10839                  The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840  iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating             Genres      Last Updated         Current Ver  \
10838       Everyone            Medical  January 20, 2017                 1.0   
10839     Mature 17+  Books & Reference  January 19, 2015  Varies with device   
10840       Everyone          Lifestyle     July 25, 2018  Varies with device   

              Android Ver  
10838          2.2 and up  
10839  Varies with device  
10840  Varies with device

In [ ]:
#Basic informations of our Dataset
df_PlayStoreData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [ ]:
df_User_Review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [ ]:
#Print all columns name
df_PlayStoreData.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [ ]:
df_User_Review.columns

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

In [ ]:
#Does the dataframe contains any Null value?
#Checking the null values in each column
df_PlayStoreData.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

In [ ]:
df_User_Review.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

**Data Preparation and Cleaning**:

**Data preparation is the process of cleaning and transforming raw data prior to processing and analysis**. It is an important step prior to processing and often involves reformatting data, making corrections to data, and the combining of data sets to enrich data. Data cleansing or data cleaning is the process of detecting and correcting (or removing) corrupt or inaccurate records from a recordset, table, or database and refers** to identifying incomplete, incorrect, inaccurate, or irrelevant parts of the data and then replacing, modifying, or deleting the dirty or coarse data.**

We saw that the dataset contains many Null or missing values.In df_Play_store_Data **The column Rating, Type , Content Rating , Current Ver , and Android Ver contains 1474, 1, 1, 8, and 3 missing values respectively.** and **in df_User_Review colums Translated_Review, Sentiment, Sentiment_Polarity and Sentiment_Subjectivity having 26868,26863,26863 and 26863 missing values respectively**.
Will it not be better if we can define a function to get more useful information about the different attributes of the dataset, also there is one more valid point in defining a function which it will be reusable, and we are going to utilize our defined function several times in future

We can see that we have some missing values in the dataset. Lets have a detailed look of the missing values in the dataframe.

Let's define a function to view some more insights of the dataset

In [ ]:
def printinfo():
    temp = pd.DataFrame(index=df_PlayStoreData.columns)
    temp['data_type'] = df_PlayStoreData.dtypes
    temp['null_count'] = df_PlayStoreData.isnull().sum()
    temp['unique_count'] = df_PlayStoreData.nunique()
    return temp

printinfo()

data_type  null_count  unique_count
App               object           0          9660
Category          object           0            34
Rating           float64        1474            40
Reviews           object           0          6002
Size              object           0           462
Installs          object           0            22
Type              object           1             3
Price             object           0            93
Content Rating    object           1             6
Genres            object           0           120
Last Updated      object           0          1378
Current Ver       object           8          2832
Android Ver       object           3            33

In [ ]:
def printinfor():
    temp = pd.DataFrame(index=df_User_Review.columns)
    temp['data_type'] = df_User_Review.dtypes
    temp['null_count'] = df_User_Review.isnull().sum()
    temp['unique_count'] = df_User_Review.nunique()
    return temp


In [ ]:
printinfor()

data_type  null_count  unique_count
App                       object           0          1074
Translated_Review         object       26868         27994
Sentiment                 object       26863             3
Sentiment_Polarity       float64       26863          5410
Sentiment_Subjectivity   float64       26863          4474

In [ ]:
#Let's have a look at the rows of the column Rating having NULL values
df_PlayStoreData[df_PlayStoreData.Rating.isnull()]

App             Category  Rating  \
23                  Mcqueen Coloring pages       ART_AND_DESIGN     NaN   
113              Wrinkles and rejuvenation               BEAUTY     NaN   
123                 Manicure - nail design               BEAUTY     NaN   
126           Skin Care and Natural Beauty               BEAUTY     NaN   
129    Secrets of beauty, youth and health               BEAUTY     NaN   
...                                    ...                  ...     ...   
10824                            Cardio-FR              MEDICAL     NaN   
10825                   Naruto & Boruto FR               SOCIAL     NaN   
10831             payermonstationnement.fr  MAPS_AND_NAVIGATION     NaN   
10835                             FR Forms             BUSINESS     NaN   
10838               Parkinson Exercices FR              MEDICAL     NaN   

      Reviews  Size  Installs  Type Price Content Rating  \
23         61  7.0M  100,000+  Free     0       Everyone   
113       182  5.7M  100,000+  Free     0   Everyone 10+   
123       119  3.7M   50,000+  Free     0       Everyone   
126       654  7.4M  100,000+  Free     0           Teen   
129        77  2.9M   10,000+  Free     0     Mature 17+   
...       ...   ...       ...   ...   ...            ...   
10824      67   82M   10,000+  Free     0       Everyone   
10825       7  7.7M      100+  Free     0           Teen   
10831      38  9.8M    5,000+  Free     0       Everyone   
10835       0  9.6M       10+  Free     0       Everyone   
10838       3  9.5M    1,000+  Free     0       Everyone   

                                Genres        Last Updated Current Ver  \
23     Art & Design;Action & Adventure       March 7, 2018       1.0.0   
113                             Beauty  September 20, 2017         8.0   
123                             Beauty       July 23, 2018         1.3   
126                             Beauty       July 17, 2018        1.15   
129                             Beauty      August 8, 2017         2.0   
...                                ...                 ...         ...   
10824                          Medical       July 31, 2018       2.2.2   
10825                           Social    February 2, 2018         1.0   
10831                Maps & Navigation       June 13, 2018   2.0.148.0   
10835                         Business  September 29, 2016       1.1.5   
10838                          Medical    January 20, 2017         1.0   

      Android Ver  
23     4.1 and up  
113    3.0 and up  
123    4.1 and up  
126    4.1 and up  
129    2.3 and up  
...           ...  
10824  4.4 and up  
10825  4.0 and up  
10831  4.0 and up  
10835  4.0 and up  
10838  2.2 and up  

[1474 rows x 13 columns]

We have some useful information about the dataset. i.e., we can now see the missing number of values of any attribute, its unique count, and its respective data types.
Now we can start the process of data cleaning, lets start with the column Type

In [ ]:
#Let's have a look at the rows of the column Type having NULL values
df_PlayStoreData[df_PlayStoreData.Type.isnull()]

App Category  Rating Reviews                Size  \
9148  Command & Conquer: Rivals   FAMILY     NaN       0  Varies with device   

     Installs Type Price Content Rating    Genres   Last Updated  \
9148        0  NaN     0   Everyone 10+  Strategy  June 28, 2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

Since there is only one missing value in this column, So, let’s fill the missing value. After cross-checking in the play store the missing value is found to be Free, So now we can fill the missing value with free .

In [ ]:
df_PlayStoreData['Type'].fillna("Free", inplace = True)

After filling the value we can check and see if that has been correctly placed. Now there is no null values in column Type

In [ ]:

df_PlayStoreData.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

Now let's move to column Translated_Review of Dataframe df_User_Review 

In [ ]:
df_User_Review[df_User_Review['Translated_Review'].isnull()]

App Translated_Review Sentiment  \
2            10 Best Foods for You               NaN       NaN   
7            10 Best Foods for You               NaN       NaN   
15           10 Best Foods for You               NaN       NaN   
102          10 Best Foods for You               NaN       NaN   
107          10 Best Foods for You               NaN       NaN   
...                            ...               ...       ...   
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
2                     NaN                     NaN  
7                     NaN                     NaN  
15                    NaN                     NaN  
102                   NaN                     NaN  
107                   NaN                     NaN  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[26868 rows x 5 columns]

In [ ]:
df_User_Review[df_User_Review['Sentiment'].isnull()]

App Translated_Review Sentiment  \
2            10 Best Foods for You               NaN       NaN   
7            10 Best Foods for You               NaN       NaN   
15           10 Best Foods for You               NaN       NaN   
102          10 Best Foods for You               NaN       NaN   
107          10 Best Foods for You               NaN       NaN   
...                            ...               ...       ...   
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
2                     NaN                     NaN  
7                     NaN                     NaN  
15                    NaN                     NaN  
102                   NaN                     NaN  
107                   NaN                     NaN  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[26863 rows x 5 columns]

In [ ]:
df_User_Review[df_User_Review['Sentiment_Polarity'].isnull()]

App Translated_Review Sentiment  \
2            10 Best Foods for You               NaN       NaN   
7            10 Best Foods for You               NaN       NaN   
15           10 Best Foods for You               NaN       NaN   
102          10 Best Foods for You               NaN       NaN   
107          10 Best Foods for You               NaN       NaN   
...                            ...               ...       ...   
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
2                     NaN                     NaN  
7                     NaN                     NaN  
15                    NaN                     NaN  
102                   NaN                     NaN  
107                   NaN                     NaN  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[26863 rows x 5 columns]

In [ ]:
df_User_Review[df_User_Review['Sentiment_Subjectivity'].isnull()]

App Translated_Review Sentiment  \
2            10 Best Foods for You               NaN       NaN   
7            10 Best Foods for You               NaN       NaN   
15           10 Best Foods for You               NaN       NaN   
102          10 Best Foods for You               NaN       NaN   
107          10 Best Foods for You               NaN       NaN   
...                            ...               ...       ...   
64290  Houzz Interior Design Ideas               NaN       NaN   
64291  Houzz Interior Design Ideas               NaN       NaN   
64292  Houzz Interior Design Ideas               NaN       NaN   
64293  Houzz Interior Design Ideas               NaN       NaN   
64294  Houzz Interior Design Ideas               NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
2                     NaN                     NaN  
7                     NaN                     NaN  
15                    NaN                     NaN  
102                   NaN                     NaN  
107                   NaN                     NaN  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[26863 rows x 5 columns]

**So we can see in all columns of DatFrame df_User_Review the NaN values are in same Rows and in the number of Nan are same in all columns** 
now Replacing the missing values with the Modevalue of all four columns which are having Nan/Missing values.

In [ ]:
modeValueTranslated_Review = df_User_Review['Translated_Review'].mode()

In [ ]:
df_User_Review['Translated_Review'].fillna(value=modeValueTranslated_Review[0], inplace = True)

In [ ]:
modeValueSentiment = df_User_Review['Sentiment'].mode()

In [ ]:
df_User_Review['Sentiment'].fillna(value=modeValueSentiment[0], inplace = True)

In [ ]:
modeValueSentiment_Polarity = df_User_Review['Sentiment_Polarity'].mode()

In [ ]:
df_User_Review['Sentiment_Polarity'].fillna(value=modeValueSentiment_Polarity[0], inplace = True)

In [ ]:
modeValueSentiment_Subjectivity = df_User_Review['Sentiment_Subjectivity'].mode()

In [ ]:
df_User_Review['Sentiment_Subjectivity'].fillna(value=modeValueSentiment_Subjectivity[0], inplace = True)

In [ ]:
printinfor()

data_type  null_count  unique_count
App                       object           0          1074
Translated_Review         object           0         27994
Sentiment                 object           0             3
Sentiment_Polarity       float64           0          5410
Sentiment_Subjectivity   float64           0          4474

**Now we can see after replacing the Nan with ModeValues there is no NaN is available now in User_Review DataFrame**

In [ ]:
df_User_Review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       64295 non-null  object 
 2   Sentiment               64295 non-null  object 
 3   Sentiment_Polarity      64295 non-null  float64
 4   Sentiment_Subjectivity  64295 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


Now, we can move on to the column Content Rating  of df_PlayStoreData 

In [ ]:
df_PlayStoreData[df_PlayStoreData['Content Rating'].isnull()]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

By, looking only at these rows it is not easy to say what's actually missing in this row. let us have a look at all of its near rows data. For this purpose, we have iloc and loc function.

In [ ]:
df_PlayStoreData.loc[10468:10477, :]

App         Category  Rating  \
10468                         Tassa.fi Finland        LIFESTYLE     3.6   
10469              TownWiFi | Wi-Fi Everywhere    COMMUNICATION     3.9   
10470                               Jazz Wi-Fi    COMMUNICATION     3.4   
10471                         Xposed Wi-Fi-Pwd  PERSONALIZATION     3.5   
10472  Life Made WI-Fi Touchscreen Photo Frame              1.9    19.0   
10473                  osmino Wi-Fi: free WiFi            TOOLS     4.2   
10474                             Sat-Fi Voice    COMMUNICATION     3.4   
10475                         Wi-Fi Visualizer            TOOLS     3.9   
10476                    Lennox iComfort Wi-Fi        LIFESTYLE     3.0   
10477              Sci-Fi Sounds and Ringtones  PERSONALIZATION     3.6   

      Reviews    Size     Installs  Type     Price Content Rating  \
10468     346    7.5M      50,000+  Free         0       Everyone   
10469    2372     58M     500,000+  Free         0       Everyone   
10470      49    4.0M      10,000+  Free         0       Everyone   
10471    1042    404k     100,000+  Free         0       Everyone   
10472    3.0M  1,000+         Free     0  Everyone            NaN   
10473  134203    4.1M  10,000,000+  Free         0       Everyone   
10474      37     14M       1,000+  Free         0       Everyone   
10475     132    2.6M      50,000+  Free         0       Everyone   
10476     552    7.6M      50,000+  Free         0       Everyone   
10477     128     11M      10,000+  Free         0       Everyone   

                  Genres        Last Updated Current Ver   Android Ver  
10468          Lifestyle        May 22, 2018         5.5    4.0 and up  
10469      Communication      August 2, 2018       4.2.1    4.2 and up  
10470      Communication   February 10, 2017         0.1    2.3 and up  
10471    Personalization      August 5, 2014       3.0.0  4.0.3 and up  
10472  February 11, 2018              1.0.19  4.0 and up           NaN  
10473              Tools      August 7, 2018     6.06.14    4.4 and up  
10474      Communication   November 21, 2014     2.2.1.5    2.2 and up  
10475              Tools        May 17, 2017       0.0.9    2.3 and up  
10476          Lifestyle      March 22, 2017      2.0.15  2.3.3 and up  
10477    Personalization  September 27, 2017         4.0    4.0 and up

We can clearly see that row 10472 has missing data for the Category column and all the prevailing column values are being replaced with its previous column. A better idea will be to drop this row from our data frame.

In [ ]:
df_PlayStoreData.dropna(subset = ['Content Rating'], inplace=True)

We are having some of the unwanted columns which will be of not much use in the analysis process. So let’s drop those columns.


In [ ]:
df_PlayStoreData.drop(['Current Ver','Last Updated', 'Android Ver'], axis=1, inplace=True)

In [ ]:
df_PlayStoreData.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10840 entries, 0 to 10840
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10840 non-null  object 
 1   Category        10840 non-null  object 
 2   Rating          9366 non-null   float64
 3   Reviews         10840 non-null  object 
 4   Size            10840 non-null  object 
 5   Installs        10840 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10840 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10840 non-null  object 
dtypes: float64(1), object(9)
memory usage: 931.6+ KB


Now, we can fix the Rating column which contains a total of 1474 of missing values. Replacing the missing values with the Modevalue of that entire column

In [ ]:
modeValueRating = df_PlayStoreData['Rating'].mode()

In [ ]:
df_PlayStoreData['Rating'].fillna(value=modeValueRating[0], inplace = True)

Finally, after fixing all the missing values, we should have a look at our data frame, We defined a function as printinfo() . So, it’s time to use that function.

In [ ]:
printinfo()

data_type  null_count  unique_count
App               object           0          9659
Category          object           0            33
Rating           float64           0            39
Reviews           object           0          6001
Size              object           0           461
Installs          object           0            21
Type              object           0             2
Price             object           0            92
Content Rating    object           0             6
Genres            object           0           119

All the columns have the null_count as zero, which indicates that now the data frame doesn’t contain any missing values.

Now we are done with the data cleansing part and in a state to start the work for data preparation.

Columns like Reviews, Size, Installs, & price should have an intor floatdatatype, But here we can see of objecttype, So let’s convert them to their respective correct type.
Starting with the column Reviews , converting its type to int .

In [ ]:
df_PlayStoreData['Reviews'] = df_PlayStoreData.Reviews.astype(int)
df_PlayStoreData['Reviews'].dtype

dtype('int64')

Now, the reviews column has been converted to int type, so now we can move to the Column: Size

Converting the Size Column from object to integer, but this column contains some of the special characters **like , , + , M , K** & also it has a some of the value as Varies with device . **We need to remove all of these** and then convert it to int or float .

# Removing the +Symbol:

In [ ]:
df_PlayStoreData['Size'] = df_PlayStoreData.Size.apply(lambda x: x.strip('+')) # Removing the + Sign

# **Now Removing ,Symbol:**

In [ ]:
df_PlayStoreData['Size'] = df_PlayStoreData.Size.apply(lambda x: x.replace(',', '')) # For removing the `,`

# **Replacing the M symbol by multiplying the value with 1000000:**


In [ ]:
df_PlayStoreData['Size'] = df_PlayStoreData.Size.apply(lambda x: x.replace('M', 'e+6')) # For converting the M to Mega


## **Replacing the k by multiplying the value with 1000:**

In [ ]:
df_PlayStoreData['Size'] = df_PlayStoreData.Size.apply(lambda x: x.replace('M', 'e+6')) # For converting the M to Mega


# Replacing the Varies with device value with Nan :

In [ ]:
df_PlayStoreData['Size'] = df_PlayStoreData.Size.replace('Varies with device', np.NaN)

# Now, finally converting all these values to numeric type:

In [ ]:
df_PlayStoreData['Size'] = pd.to_numeric(df_PlayStoreData['Size'])  # Converting the string to Numeric type

ValueError: ignored

So, after performing all of these operations, we should have a detailed look at that column

In [ ]:
printinfo()

Since we converted the Varies with device value to Nan , so we have to do something with those set of Nan values data. It will be a better idea to drop the Rows of the column Size having Nanvalues because it will be not an efficient idea to replace those values with mean or mode since the size of some apps would be too large and some of them too small.

In [ ]:
df_PlayStoreData.dropna(subset = ['Size'], inplace=True)

# Column: Installs :
To convert this column from object to integer type. First of all, we will need to remove the +symbol from these values.

In [ ]:
df_PlayStoreData['Installs'] = df_PlayStoreData.Installs.apply(lambda x: x.strip('+'))

and then let’s remove the , symbol from the numbers.

In [ ]:
df_PlayStoreData['Installs'] = df_PlayStoreData.Installs.apply(lambda x: x.replace(',', ''))

Lastly, we can now convert it from string type to numeric type, and then have a look at our dataset.

In [ ]:
df_PlayStoreData['Installs'] = pd.to_numeric(df_PlayStoreData['Installs'])

In [ ]:
printinfo()

So, now we are only left with the Price column.
Column: Price :
Converting this column from objectto Numeric type.

In [ ]:
df_PlayStoreData['Price'].value_counts()

The values contain a special symbol $ which can be removed and then converted to the numeric type.

In [ ]:
df_PlayStoreData['Price'] = df_PlayStoreData.Price.apply(lambda x: x.strip('$'))

In [ ]:
df_PlayStoreData['Price'] = pd.to_numeric(df_PlayStoreData['Price'])

After fixing all the issues, we should have a final look at the data frame.

In [ ]:
printinfo()

Now, we are finally done. In this section Data Preparation and Cleaning. We can see that the original dataset contained 10841 Rows and 13 Columns. It contained App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Curernt Ver, and Android VerColumns. But after cleansing the dataset and dropping the unwanted rows and columns having Null Values and Garbage data from the data frame, we are left with 9145 Rows and 10 Columns.

In [ ]:
df_PlayStoreData.info()

In [ ]:
df_PlayStoreData.shape

# Exploratory Analysis and Visualization
In statistics, exploratory data analysis is an approach to analyzing data sets to summarize their main characteristics, often with visual methods. A statistical model can be used or not, but primarily EDA is for seeing what the data can tell us beyond the formal modeling or hypothesis testing task. Data visualization is the graphic representation of data. It involves producing images that communicate relationships among the represented data to viewers of the images. This communication is achieved through the use of a systematic mapping between graphic marks and data values in the creation of the visualization. This mapping establishes how data values will be represented visually, determining how and to what extent the property of a graphic mark, such as size or color, will change to reflect changes in the value of a datum.
Let’s begin by importing matplotlib.pyplot and seaborn , and at the same time set our fig size, font size, etc.

In [ ]:
import seaborn as sns
import matplotlib
import matplotlib.pyplot as plt
%matplotlib inline

sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 14
matplotlib.rcParams['figure.figsize'] = (9, 5)
matplotlib.rcParams['figure.facecolor'] = '#00000000'

Now it is time to unveil the real strength of data analysis, i.e., to get an insight, and learn the trend, pattern and get answers to some of the questions related to the dataset.

**Can we see what are the top categories in the play store, which contains the highest number of apps? Well, let us try to.**

In [ ]:
y = df_PlayStoreData['Category'].value_counts().index
x = df_PlayStoreData['Category'].value_counts()
xsis = []
ysis = []
for i in range(len(x)):
    xsis.append(x[i])
    ysis.append(y[i])

We have defined our x and y axis. Let us plot and see:-

In [ ]:
plt.figure(figsize=(18,13))
plt.xlabel("Count")
plt.ylabel("Category")

graph = sns.barplot(x = xsis, y = ysis, palette= "husl")
graph.set_title("Top categories on Google Playstore", fontsize = 25);

So there are all total of 33 categories in the dataset from the above output we can come to the conclusion that in the play store most of the apps are under Family & Game category and least are of Beauty & Comics Category.

Which category of Apps from the ‘Content Rating’ column is found more on the play store?

In [ ]:
x2 = df_PlayStoreData['Content Rating'].value_counts().index
y2 = df_PlayStoreData['Content Rating'].value_counts()

x2sis = []
y2sis = []
for i in range(len(x2)):
    x2sis.append(x2[i])
    y2sis.append(y2[i])

In [ ]:
plt.figure(figsize=(12,10))
plt.bar(x2sis,y2sis,width=0.8,color=['#15244C','#FFFF48','#292734','#EF2920','#CD202D','#ECC5F2'], alpha=0.8);
plt.title('Content Rating',size = 20);
plt.ylabel('Apps(Count)');
plt.xlabel('Content Rating');

From the above plot, we can see that the Everyone category has the highest number of apps.

Let’s have a look at the distribution of the ratings of the data frame.

In [ ]:
plt.figure(figsize=(15,9))
plt.xlabel("Rating")
plt.ylabel("Frequency")
graph = sns.kdeplot(df_PlayStoreData.Rating, color="Blue", shade = True)
plt.title('Distribution of Rating',size = 20);

From the above graph, we can come to the conclusion that most of the apps in the google play store are rated between 3.5 to 4.8.


**Let’s plot a visualization graph to view what portion of the apps in the play store are paid and free.**

In [ ]:
plt.figure(figsize=(10,10))
labels = df_PlayStoreData['Type'].value_counts(sort = True).index
sizes = df_PlayStoreData['Type'].value_counts(sort = True)
colors = ["blue","lightgreen"]
explode = (0.2,0)
plt.pie(sizes, explode=explode, labels=labels, colors=colors, autopct='%1.1f%%', shadow=True, startangle=0)
plt.title('Percent of Free Vs Paid Apps in store',size = 20)
plt.show()

**From the above graph, we can see that 92%(Approx.) of apps in the google play store are free and 8%(Approx.) are paid**

Which category App’s have the most number of installs?
To answer this question we need to create a separate data frame out of **our df_PlayStoreData data frame which will contain a grouped value by Category and Installs .**

In [ ]:
highest_Installs_df = df_PlayStoreData.groupby('Category')[['Installs']].sum().sort_values(by='Installs', ascending=False)

**Now, let us plot it out:**

In [ ]:
x2sis = []
y2sis = []

for i in range(len(highest_Installs_df)):
    x2sis.append(highest_Installs_df.Installs[i])
    y2sis.append(highest_Installs_df.index[i])

plt.figure(figsize=(18,13))

plt.xlabel("Installs")
plt.ylabel("Category")
graph = sns.barplot(x = x2sis, y = y2sis, alpha =0.9, palette= "viridis")
graph.set_title("Installs", fontsize = 25);

From the above visualization, it can be interpreted that the top categories with the highest installs are Game, Family, Communication, News & Magazines, & Tools.

We have done somewhat a good number of exploratory data analysis till now and in a state to finally answer some of the most common and in-demand questions which every App developer or any business company will love to know.

Asking and Answering Questions
With the help of Data Analysis, we can answer many questions that can’t be answered just by looking at the dataset. By querying on any dataset, and understanding the pattern and rate of growth and fall of any values we can come to many conclusions, and get insightful information from it. So let's start

What are the Top 10 installed apps in any category?
So, we have to be able to answer this not only for a single category but for many, i.e., we will need to define a function which should be able to return us a nice plot for any Category the name provided by any user as an argument to it.

In [ ]:
def findtop10incategory(str):
    str = str.upper()
    top10 = df_PlayStoreData[df_PlayStoreData['Category'] == str]
    top10apps = top10.sort_values(by='Installs', ascending=False).head(10)
    # Top_Apps_in_art_and_design
    plt.figure(figsize=(15,12))
    plt.title('Top 10 Installed Apps',size = 20);    
    graph = sns.barplot(x = top10apps.App, y = top10apps.Installs)
    graph.set_xticklabels(graph.get_xticklabels(), rotation= 45, horizontalalignment='right');

After we are done with defining the function, it’s time to check and see if everything is working fine. So let’s test it by passing Sports category to the above-defined function.

In [ ]:
findtop10incategory('Sports')

From the above graph, we can see that in the Sports category FIFA Soccer, and Dream League Soccer 2018 has the highest installs. In the same way by passing different category names to the function, we can get the top 10 installed apps.

Which are the top 10 expensive Apps in the play store?
We will again need to create a separate data frame.

In [ ]:
top10PaidApps = df_PlayStoreData[df_PlayStoreData['Type'] == 'Paid'].sort_values(by='Price', ascending=False).head(11)
top10PaidApps

From the above data frame, we will need to drop an app name, because its’ name will be creating a mess in the plot.

In [ ]:
top10PaidApps_df = top10PaidApps[['App', 'Installs']].drop(9934)

So Finally let’s plot and visualize the top 10 paid apps on the play store.

In [ ]:
plt.figure(figsize=(15,12));
plt.pie(top10PaidApps_df.Installs, explode=None, labels=top10PaidApps_df.App, autopct='%1.1f%%', startangle=0);
plt.title('Top Expensive Apps Distribution',size = 20);
plt.legend(top10PaidApps_df.App, 
           loc="lower right",
           title="Apps",
           fontsize = "xx-small"
          );

From the above graph, we can interpret that the App I am rich is the most expensive app in the google play store followed by I am Rich Premium. we also had to drop one-row data for this visualization because the language of the app was Chinese and it was messing with the pie chart, visualization.

**Which are the Apps with the highest number of reviews?**

In [ ]:
Apps_with_Highest_rev =df_PlayStoreData.sort_values(by='Reviews', ascending=False).head(20)

In [ ]:
Apps_with_Highest_rev


From the above data frame we can interpret, and come to the conclusion that the Apps like Clash of Clans, Subway Surfers, Clash Royale, and Candy Crush Saga has the highest number of reviews on google play store.

**What are the count of Apps in different genres?**
By creating a data frame, let’s define our x and y axis, which will be required for plotting the graph.

In [ ]:
topAppsinGenres = df_PlayStoreData['Genres'].value_counts().head(50)

In [ ]:
topAppsinGenres

In [ ]:
x3sis = []
y3sis = []

for i in range(len(topAppsinGenres)):
    x3sis.append(topAppsinGenres.index[i])
    y3sis.append(topAppsinGenres[i])

Finally, we are in a state to plot and gain an insight into our raised question.

In [ ]:
plt.figure(figsize=(15,9))
plt.ylabel('Genres(App Count)')
plt.xlabel('Genres')
graph = sns.barplot(x=x3sis,y=y3sis,palette="deep")
graph.set_xticklabels(graph.get_xticklabels(), rotation=90, fontsize=12)
graph.set_title("Top Genres in the Playstore", fontsize = 20);

**From the above visualization, we can see that the Highest Number of Apps found in the Tools and Entertainment genres followed by Education, Medical and many more.**

**Which are the apps that have made the highest-earning?**

For answering these questions we will need to perform some extra operation to the data frame, i.e., we will need to create a separate data frame, and then multiply the Price column and the Installs column in order to get the earning of any particular app. So, let's start the process.

In [ ]:
Paid_Apps_df =df_PlayStoreData[df_PlayStoreData['Type'] == 'Paid']

Now from the above data frame, we will need to separate out the columns which we will require.

In [ ]:
earning_df = Paid_Apps_df[['App', 'Installs', 'Price']]


We can now add a separate column Earnings to our new data frame which we will create by multiplying the two-column Price and Installs .

In [ ]:
earning_df['Earnings'] = earning_df['Installs'] * earning_df['Price'];

Now let us sort the above data by Earnings and Price .

In [ ]:
earning_df_sorted_by_Earnings = earning_df.sort_values(by='Earnings', ascending=False).head(50)

In [ ]:
earning_df_sorted_by_Price = earning_df_sorted_by_Earnings.sort_values(by='Price', ascending=False)

Finally, we can plot the graph and find out which are the apps with the highest number of earnings.

In [ ]:
# PLot a bar chart of earning at y and app names at x
plt.figure(figsize=(15,9))
plt.bar(earning_df_sorted_by_Price.App, earning_df_sorted_by_Price.Earnings, width=1.1, label=earning_df_sorted_by_Price.Earnings)
plt.xlabel("Apps")
plt.ylabel("Earnings")
plt.tick_params(rotation=90)
plt.title("Top Earning Apps");

The top five apps with the highest earnings found on google play store are:-

.I am Rich

.I am Rich Premium

.Hitman Sniper

.Grand Theft Auto: San Andreas

.Facetune - For Free

We have finally come to an end to our analysis, and hope that if you have reached till here it must have been interesting or useful to you.

**Inferences and Conclusion**

After Analyzing the dataset we have got answers to some of the serious & interesting questions which any of the android users would love to know.

so after analyzing the Dataset we conclude that there are all total of 33 categories in the dataset in the play store most of the apps are under Family & Game category and least are of Beauty & Comics Category.

In content Rating column we can see that the Everyone category has the highest number of apps.

If we talk about Rating Column we can come to the conclusion that most of the apps in the google play store are rated between 3.5 to 4.8.

Refer to Type column we can see that 92%(Approx.) of apps in the google play store are free and 8%(Approx.) are paid.

From the column Intall visualization, it can be interpreted that the top categories with the highest installs are Game, Family, Communication, News & Magazines, & Tools.

From visualization graph, we can see that in the Sports category FIFA Soccer, and Dream League Soccer 2018 has the highest installs. In the same way by passing different category names to the function, we can get the top 10 installed apps.

We also saw that top expensive/Top earning App I am rich is the most expensive app in the google play store followed by I am Rich Premium. 

 we also saw that in colum Genres Highest Number of Apps found in the Tools and Entertainment genres followed by Education, Medical and many more.


#First EDA project we learnt much more things 